storybookjs/storybook
# Addon-docs: Fix Memo React components in ArgsTable

Issues: #8894 #9586 

Storybook seems to be missing correct exotic component type extraction.
For Forward Ref, if the wrapped component has `propTypes` defined, it will work, but if it doesn't while the inner render fn does, it won't work.
For Memo, it doesn't work at all.

## What I did

- [x] Created React lib helpers for `isMemo` & `isForwardRef` (cannot use `react-is` counterpart as they always expect Elements not Components)
- [x] Refactored `jsxDecorator` to use new helpers & fetch `.__docgen.displayName` if available!
- [x] Fixed `extractProps` so it can get types from Memo render fn
- [x] Consolidated example exotic components
- [x] Updated stories to reflect new behaviour

**NOTE**: Unfortunately, docgen doesn't work properly if you try to attach `propTypes` to the wrapped Memo component, it can only work when the Memo's inner/render fn itself has `propTypes` attached.

## How to test

- Is this testable with Jest or Chromatic screenshots?
No?
- Does this need a new example in the kitchen sink apps?
Yes, added new stories on `official-storybook` app.
- Does this need an update to the documentation?
No.

If your answer is yes to any of these, please make sure to include it in your PR.
